Fix/disclaimer update

workerbee22 opened the pull request:
## Changes

- Fix app startup to display the disclaimer on first run only or if the version has changed from the version the user accepted
- Fix back button so it cannot be used to skip the disclaimer screen
- Fix bug where home screen may be shown twice or Back button twice needed to exit the app on Android. iOS does not permit an app to exit itself.
- Fix bug where we get Overflow errors on some lower density device screens
- Add versioning within the app, so if the version flag persisted to device storage changes compared to the current app release, then reprompt user to accept the disclaimer
- New version of the Disclaimer page content
- Added scrollbar on touch when scrolling
- Added button fixed at bottom of the screen rather than have to scroll to bottom to press
- SafeArea for disclaimer so does not impact system UI on Android or iOS
- Information screen disclaimer shows that it was accepted by the user, version and date & time user agreed to the disclaimer

## Things to note

There are also some general app improvements like:
- strings.dart file for fixed strings used in code
- changes to some routes to fix the double back issue
- other general cleanup required.

workerbee22 committed:
Revised disclaimer screen to allow scrolling on any screen size.
Reworked app startup logic to remove issues when could not exit the app and/or could skip disclaimer without accepting it.
Fixed issue where after accepting the disclaimer could use back key to return to the disclaimer screen.
Added Decline button for Android only (iOS forbids exiting app from within an app)
Still need to improve disclaimer screen now that functional changes made.
Other architectural cleanups for routes and navigator.

workerbee22 committed:
Merge remote-tracking branch 'origin/fix/disclaimer-update' into fix/disclaimer-update

# Conflicts:
#	lib/hard_data.dart

workerbee22 committed:
Revised disclaimer screen to allow scrolling on any screen size.
Reworked app startup logic to remove issues when could not exit the app and/or could skip disclaimer without accepting it.
Fixed issue where after accepting the disclaimer could use back key to return to the disclaimer screen.
Added Decline button for Android only (iOS forbids exiting app from within an app)
Still need to improve disclaimer screen now that functional changes made.
Other architectural cleanups for routes and navigator.

workerbee22 committed:
Revised disclaimer screen to allow scrolling on any screen size.
Reworked app startup logic to remove issues when could not exit the app and/or could skip disclaimer without accepting it.
Fixed issue where after accepting the disclaimer could use back key to return to the disclaimer screen.
Added disclaimer versioning. When the version is updated the user will be re-asked to accept the disclaimer.
Added SafeArea to be clear of system ui.
Updated disclaimer screen in Information to show the user they accpted the current version, the version flag and the date/time they agreed.
Still need to improve disclaimer screen now that functional changes made.
Other architectural cleanups for routes and navigator.

workerbee22 commented:
This isn't super tight code, but it is quite urgent (apologies been doing CI & CD setup). So keen to merge this and then do a separate PR for making it look nicer.

workerbee22 committed:
Merge branch 'master' into fix/disclaimer-update

# Conflicts:
#	lib/intro_router.dart
#	lib/main.dart
#	lib/routes.dart
#	lib/view/disclaimer_view.dart
#	lib/view/info_view.dart
#	lib/view/intubation/algorithm/intubation_algorithm_page.dart
#	lib/view/intubation/checklist/intubation_checklist_page.dart
#	lib/view/intubation/guidance/intubation_guidance_page.dart
#	lib/view/ppe/ppe_off_guidance_method_1.dart
#	lib/view/ppe/ppe_off_guidance_method_2.dart
#	lib/view/staff_welfare/your_welfare_view.dart
#	lib/widget/cards/ppe_step_card.dart
#	pubspec.yaml

workerbee22 committed:
Updated this branch with master branch updates. All merged.
This branch then also tested on Android and iOS.

workerbee22 commented:
Ok just pushed a new commit that has caught up with master branch. 

Note many file changes are due only to:
- Strings class in `strings.dart` file
- Styles class renamed from AppStyles previously

workerbee22 committed:
Fixed lint warnings causing CI builds to fail.

sallychung commented:
@workerbee22 I noticed that the if I access the Disclaimer from the Info section, there’s no back button so I get stuck. Would you like me to raise an issue for this as a separate task?

iPhone 6s
![image](https://user-images.githubusercontent.com/12321678/78235565-8a6ea180-7524-11ea-9913-6bba905217e8.png)

workerbee22 committed:
Fixed test 'show home screen if user has previously accepted disclaimer' by updating test to look for the correct value.

workerbee22 commented:
@sallychung Yep its a bug.

workerbee22 commented:
So I'm working on these before finalise this PR:

- fixing the bug about no back button on ios
- fixing tests for updated shared preferences values and adding new tests

alessandrofavero commented on lib/view/disclaimer_view.dart:
```diff
@@ -1,206 +1,202 @@
-    with SingleTickerProviderStateMixin {
-  final String _title = 'Disclaimer and conditions of use';
+class _DisclaimerViewState extends State<DisclaimerView> {
+  @override
+  void initState() {
+    super.initState();
+  }
```
If you are not doing anything in the initState(), I would remove it.

workerbee22 replied:
Good point ... I was playing with something in there. I usually keep initState and dispose in there regardless .... but thats just me. I'll remove 👍

alessandrofavero commented on lib/view/disclaimer_view.dart:
```diff
@@ -1,206 +1,202 @@
-      //textAlign: TextAlign.justify,
-    ),
-  );
+  @override
+  void dispose() {
+    super.dispose();
+  }
```
Same as per initState(), more compact code.

workerbee22 replied:
ditto 👍

alessandrofavero commented on lib/view/disclaimer_view.dart:
```diff
@@ -1,206 +1,202 @@
-    final _agreeButton = Container(
-      margin: const EdgeInsets.only(left: 16, right: 16, bottom: 16),
+    /// Agree button - button a user taps to agree to the disclaimer
+    final Widget _agreeButton = Container(
```
I would convert this to a function like _agreeButton(BuildContext context) and move it outside the build method.

workerbee22 replied:
Good idea 👍

alessandrofavero commented on lib/view/disclaimer_view.dart:
```diff
@@ -1,206 +1,202 @@
-                                  _agreedText,
-                              ],
+    // Content of the page
+    Widget _agreedMessage(dynamic version, dynamic dateStampString) {
```
I would move this function outside the build method

workerbee22 replied:
Good idea 👍

alessandrofavero commented on lib/view/disclaimer_view.dart:
```diff
@@ -1,206 +1,202 @@
+      );
+    }
+
+    Widget _disclaimerScrollingContent(BuildContext context) {
```
I would move this function outside the build method

workerbee22 replied:
ditto 👍

alessandrofavero commented on lib/view/disclaimer_view.dart:
```diff
@@ -1,206 +1,202 @@
+          backgroundColor: Colors.white,
+          elevation: 4.0,
+          iconTheme: Styles.appBarIconTheme,
+          automaticallyImplyLeading: false,
```
You need to implement a check here otherwise the user will not be able to go back if they check the disclaimer from the info page.

workerbee22 replied:
Yeah thats the bug in the Information screen view. Need to move the FutureBuilder up to get that value in here.

workerbee22 replied:
I'm still torn on the Disclaimer view at app startup vs Disclaimer view from the Information screen, because the only thing they really have in common is the disclaimer content itself. But happy to keep your original approach of using it for both for now, unless they change more in the future and get even more different 👍

workerbee22 replied:
All done.

alessandrofavero reviewed (commented):
Hey @workerbee22 , there are some minor changes that I would address and a check implementation in the automaticallyImplyLeading. Other than that LGTM great job 👍👍

workerbee22 commented:
Thanks @alessandrofavero ... good ideas. I also wanted to keep your original approach as much as possible. Back to desk now and working on these.

Also have some tests to fix and additional tests, but want this in in next few hours 👍

workerbee22 committed:
Fixed bug where no back button shown when displaying Disclaimer from the Information screen.
Fixed disclaimer version update logic bug.
Fixed bug where at app startup, 'I agree' button was not show.
Tests workaround for now, as several more to be added, but known issue with > 1 SharedPreferences tests.
Fixed MyApp const keyword missing.
Updated code layout as per suggestions.
General code cleanup.

workerbee22 commented:
@alessandrofavero 
All raised issues resolved and changes made, can you perhaps review if you get a chance because you know this code best. Then I can merged.

workerbee22 commented:
@maks @ericjiang97 @rockgecko-development If one of you could also review during the day? We can finally get this in ... pretty later.

alessandrofavero commented on lib/view/disclaimer_view.dart:
```diff
@@ -37,6 +28,7 @@ class _DisclaimerViewState extends State<DisclaimerView> {
 
     disclaimerValues.agreed = prefs.getBool(Strings.settingDisclaimerAgreed) ?? false;
+    // Disclaimer version flag starts at '1' normally
     disclaimerValues.version = prefs.getString(Strings.settingDisclaimerVersion) ?? '0';
```
cascade?
final disclaimerValues = DisclaimerDetails()
..agreed = .......
..version = ....... ;

workerbee22 replied:
Given Disclaimer is all tied up with app startup behaviour, I thought I'd make the code a little more 'readable' vs less lines of code. We can definitely tighten up, but with many new Flutter devs on the project I'm err-ing on the side of 'readability' an making the code clean and simple.

workerbee22 replied:
But we should discuss this sort of stuff in the nightly meetings or slack. Hard to know what everyone would prefer?

alessandrofavero reviewed (approved):
LGTM 👍Tested the codemagic build on Pixel4XL and it works great.

workerbee22 commented:
@alessandrofavero Wrote the original code and knows the features around Disclaimer really well. I've also tested on 4+ devices iOS and Android, so going to merge it.

workerbee22 commented:
All good ... merged into master, built, published and tested on Android & iOS 👍
